### ISSUE TITLE
frustration with inline script

### ISSUE THREAD

**Issue by jbaumflek** [dotnet/AspNetCore.Docs#34425]
### Description

Perhaps this is not a documentation issue, or rather, this document in particular.
As we discovered in pen testing of our Blazor server-side app, we were missing CSP headers.  As I put them in, I am running into a BIG problem because, as much of the Blazor sample documentation indicates, I used inline script all over the place.  If I have to hash every single place in the app where there is code, I can't imagine the size of the csp header for all the hashes I'll need.
I must be missing something...or there is a fundamental flaw in Blazor/razor re: security and XSS.

[Enter feedback here]

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/blazor/security/content-security-policy?view=aspnetcore-8.0#server-side-blazor-apps

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/security/content-security-policy.md

### Document ID

6e0b5c52-90a1-5ca6-bfad-df33a8beae6c

### Article author

@guardrex

[Related Issues](https://github.com/dotnet/AspNetCore.Docs/issues?q=is%3Aissue+is%3Aopen+6e0b5c52-90a1-5ca6-bfad-df33a8beae6c)

**Comment by guardrex** [comment]
Hello @jbaumflek ... 

> there is a fundamental flaw in Blazor/razor re: security and XSS

No, it's just not the intended best case scenario to make liberal use of inline scripts in Blazor. If you want to do that, I think you'd probably use the nonce-source approach. Could you look into that approach and see if it would aid you? Scroll down in this section ...

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#unsafe_inline_script

Also, consider not using a lot of inline scripts if you can. See if you can leverage [JS collocation](https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/location-of-javascript?view=aspnetcore-9.0#load-a-script-from-an-external-javascript-file-js-collocated-with-a-component) instead.

Leave this issue open until I get back to work after New Year's. I'd like to take a closer look at it for possible coverage.

*Happy New Year!* 🎉 ... I'll get back to you probably next week or the week after.

**Comment by guardrex** [comment]
Took a closer look.

`<script>` tags can't be dynamically modified by Blazor. Therefore, the nonce-source approach is completely ruled out. This should be mentioned in the *CSP* article.

Now, might a hash-source approach work with controlling `<head>` content (i.e., dynamically create the hash in a component for the inline script and load it via controlling `<head>` content)? Yes and no ... ***mostly NO.*** If you try to generate the hash dynamically in a component for a specific inline script, it simply doesn't work, probably because the most restrictive policy is used by the browser (i.e., the app-level CSP that you set in the `App` component). One could dynamically generate the hash for the CSP in the `App` component, but that's wild hack with far more overhead than it's worth because the inline script isn't even there, as it would be in components loaded by children of the `App` component (btw tho, the approach does work here in a test that I performed).

I'm going to note that inline JavaScript requires the inline script's hash to appear in the most restrictive policy, which is typically the policy set in the `App` component. I'll recommend not using inline scripts and focusing more on the JS collocation approach. Then, I'll ask Steve to review the PR.

WRT your asking about Blazor having some kind of feature to better manage CSPs for inline scripts, you can open an issue for the product unit to ask them about it at ...

https://github.com/dotnet/aspnetcore/issues

Please add ...

```
cc: @guardrex https://github.com/dotnet/AspNetCore.Docs/issues/34425
```

... to the bottom of your opening comment so that I can follow along and possibly take some of their response into my PR on this. I'll ping u on the PR when it goes up.

I think that they're going to tell you not to rely on a lot of inline scripting with Blazor with a CSP.

**Comment by jbaumflek** [comment]
@guardrex What do you mean here:
_I'm going to note that inline JavaScript requires the inline script's hash to appear in the most restrictive policy, which is typically the policy set in the App component_
I think I'm missing something - I am setting the CSP uniformly from the program.cs file.
What would it look like in the App component?

thanks,
Jean

**Comment by guardrex** [comment]
@adegeo ... I wouldn't think that anything you've been working on here would upset projects; but for some unknown reason, I'm closing out issues without them being automatically moved to "Done" on the project. If you can't think of any reason why all-of-a-sudden this would stop, I guess I'll open an issue for the GH folks to take a look at this.

**Comment by adegeo** [comment]
@guardrex  Looking

**Comment by adegeo** [comment]
@guardrex Can you link to specific items that I can try and track? I don't think the stuff I added does things in the context of a user, such as yourself. The actions say it's done by the policy bot.

**Comment by guardrex** [comment]
I think you're fine. I think GH is having a problem. The Action associated with marking the issue "Done" seems to be a completely separate process here.

I'll keep an 👁 on this and see if I need to open a support ticket with GH.

**Comment by adegeo** [comment]
Is this the only one you saw? I just did a test issue and it seemed fine. https://github.com/dotnet/AspNetCore.Docs/issues/34495

**Comment by guardrex** [comment]
There were two issues IIRC that failed to move to "Done." This one, and there was one other one, but I don't recall which one it was. If I have to open a support ticket, I'll track that other one down.

**Comment by adegeo** [comment]
Yeah I think it's outside of the policy bot. The know that the policy bot can't receive events from the project system currently, so I can't imagine it was interfering.

**Comment by guardrex** [comment]
Thanks for checking ... sorry to have bothered you on it. I just wasn't sure at first. It's very probably GH's fault ... and probably not just here. If it's a problem, they've probably already gotten an ear-full from folks. 😆
